prog/dftb+/prg_dftb/mainio.F90

```diff
@@ -82,524 +96,1459 @@ module mainio
+  subroutine writeRealEigvecsTxtBlacs(env, denseDesc, eigvecs, fd, parallelKS, orb, over,&
+      & iNeighbor, nNeighbor, iSparseStart, img2CentCell, species, speciesName, fileName)
 
-    !> Mapping of atoms into the central cell
-    integer, intent(in) :: img2CentCell(:)
+    !> Environment settings
+    type(TEnvironment), intent(in) :: env
+
+    !> Dense matrix descriptor
+    type(TDenseDescr), intent(in) :: denseDesc
+
+    !> Square Hamiltonian (or work array)
+    real(dp), intent(in) :: eigvecs(:,:,:)
+
+    !> Fileid (file not yet opened) to use.
+    integer, intent(in) :: fd
+
+    !> K-points and spins to process
+    type(TParallelKS), intent(in) :: parallelKS
 
     !> Orbital information
     type(TOrbitals), intent(in) :: orb
 
-    !> Sparse overlap matrix
+    !> Sparse overlap
     real(dp), intent(in) :: over(:)
 
-    !> Square Hamiltonian (or work array)
-    real(dp), intent(inout) :: HSqrReal(:,:,:)
+    !> Neighbors of each atom
+    integer, intent(in) :: iNeighbor(0:,:)
 
-    !> Work array
-    real(dp), intent(inout) :: SSqrReal(:,:)
+    !> Nr. of neighbors for each atom
+    integer, intent(in) :: nNeighbor(:)
 
-    !> orbital number in each region
-    type(listIntR1), intent(inout) :: iOrbRegion
+    !> Index array for sparse matrices
+    integer, intent(in) :: iSparseStart(:,:)
 
-    !> If present, Hamiltonian(s) are fetched from this storage into HSqrReal, instead of using
-    !> whatever is already there
-    type(OFifoRealR2), intent(inout), optional :: storeEigvecs(:)
+    !> Mapping of atoms into the central cell.
+    integer, intent(in) :: img2CentCell(:)
 
-    integer, allocatable :: iOrbs(:)
-    integer :: iSpin, iSpin2, iLev, ii, nReg, dummy
-    integer :: valshape(1)
-    real(dp) :: qState
-    real(dp), allocatable :: rVecTemp(:)
-    character(lc) :: tmpStr
+    !> Species of each atom
+    integer, intent(in) :: species(:)
+
+    !> Name of each species
+    character(*), intent(in) :: speciesName(:)
+
+    !> optional alternative file prefix, to appear as "fileName".bin
+    character(len=*), intent(in), optional :: fileName
+
+    type(linecomm) :: collector
+    real(dp), allocatable :: localEigvec(:), localFrac(:)
+    real(dp), allocatable :: globalS(:,:), globalFrac(:,:)
+    integer :: nOrb, nAtom
+    integer :: iKS, iS, iGroup, iEig
+
+    nOrb = denseDesc%fullSize
+    nAtom = size(nNeighbor)
+    allocate(globalS(size(eigvecs, dim=1), size(eigvecs, dim=2)))
+    allocate(globalFrac(size(eigvecs, dim=1), size(eigvecs, dim=2)))
+    if (env%mpi%tGroupMaster) then
+      allocate(localEigvec(nOrb))
+      allocate(localFrac(nOrb))
+    end if
+
+    call collector%init(env%blacs%orbitalGrid, denseDesc%blacsOrbSqr, "c")
+    if (env%mpi%tGlobalMaster) then
+      call prepareEigvecFileTxt(fd, .false., fileName)
+    end if
+
+    masterSlave: if (env%mpi%tGlobalMaster) then
+      do iKS = 1, parallelKS%maxGroupKS
+        group: do iGroup = 0, env%mpi%nGroup - 1
```

Comment:
Might be useful to comment the communication between groups and the interaction with the iKS loop